src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -2362,148 +2392,117 @@ private void InferTypeFrom(VariableExpressionAst variableExpressionAst, List<PST
                 return;
             }
 
-            // For certain variables, we always know their type, well at least we can assume we know.
-            if (astVariablePath.IsUnqualified)
+            // Process well known variables like: $this and $true
+            if (astVariablePath.IsUnqualified
+                && astVariablePath.UserPath.EqualsOrdinalIgnoreCase(SpecialVariables.This)
+                && (_context.CurrentTypeDefinitionAst is not null || _context.CurrentThisType is not null))
             {
-                var isThis = astVariablePath.UserPath.EqualsOrdinalIgnoreCase(SpecialVariables.This);
-                if (!isThis || (_context.CurrentTypeDefinitionAst == null && _context.CurrentThisType == null))
+                // $this is special in script properties and in PowerShell classes
+                PSTypeName typeName = _context.CurrentThisType ?? new PSTypeName(_context.CurrentTypeDefinitionAst);
+                inferredTypes.Add(typeName);
+                return;
+            }
+            else
+            {
+                if (SpecialVariables.AllScopeVariables.TryGetValue(astVariablePath.UnqualifiedPath, out Type knownType))
```

Comment:
What scenario does this fix?

Reply:
I changed from UserPath to UnqualifiedPath so the lookup is done with the actual variable name even if a scope is specified. If I didn't do that it would not see `$global:home` as one of the special variables even though it is.